### ISSUE TITLE
Channel doesn't attempt to reconnect after idle time

### ISSUE THREAD

**Issue by alamaison** [grpc/grpc#8538]
We are seeing a bug (?) with gRPC (Python) on one particular link that crosses AWS VPCs.

When the channel is idle for a certain time, the next RPC on the channel fails with `StatusCode.UNAVAILABLE`.  We would expect the underlying channel to be renegotiated silently and the RPC to succeed.  This failure happens regardless of whether we use `channel_ready_future`/`cygrpc.check_connectivity_state` on the channel prior to making the RPC.

The following code reproduces the issue:

``` python
import time
import logging

from third_party.python import grpc
from service_proto import foo_pb2

target = 'server-in-other-aws-vpc:50005'
channel = grpc.insecure_channel(target)
stub = foo_pb2.FooServiceStub(channel)
NUMBER_OF_REQUEST_TRIES = 2
# This delay produces the grpc.StatusCode.UNAVAILABLE error state
RETRY_DELAY_IN_SECS = 400
# Where-as this delay works as expected
# RETRY_DELAY_IN_SECS = 300
ERROR_RETRY_DELAY_IN_SECS = 0.2

def log_connectivity_changes(connectivity):
    print("Channel changed status to %s." % connectivity)

channel.subscribe(log_connectivity_changes)

def make_request():
    previous_connectivity_state = channel._channel.check_connectivity_state(True)
    print('check_connectivity_state returned with %s.' % previous_connectivity_state)
    response = stub.GetFoo.future(foo_pb2.GetFooRequest(), 10)
    print('Got successful result: %s' % response.result())

try:
    for i in range(NUMBER_OF_REQUEST_TRIES - 1):
        print('Making request nr %s.' % (i + 1))
        make_request()
        print('Sleeping now for %s minutes and then retrying request to produce UNAVAILABLE '
              'error.' % (RETRY_DELAY_IN_SECS / 60.0))
        time.sleep(RETRY_DELAY_IN_SECS)
    print('Making final request.')
    make_request()
except grpc.RpcError as rpc_error:
    if rpc_error.code() == grpc.StatusCode.UNAVAILABLE:
        logging.exception(rpc_error)
        print('Sleeping now for %s seconds and then retrying request, which will be '
              'successful.' % ERROR_RETRY_DELAY_IN_SECS)
        time.sleep(ERROR_RETRY_DELAY_IN_SECS)
        print('Retrying request now.')
        make_request()
    else:
        raise rpc_error
```

If we are idle for 400s we get the following failure:

```
Making request nr 1.
Channel changed status to ChannelConnectivity.IDLE.
 check_connectivity_state returned with 0.
Channel changed status to ChannelConnectivity.READY.
Got successful result: error {
  code: 50001
  err_message: "Status message"
}

Sleeping now for 6.66666666667 minutes and then retrying request to produce UNAVAILABLE error.
Making final request.
check_connectivity_state returned with 2.
Channel changed status to ChannelConnectivity.IDLE.
ERROR:root:<_Rendezvous of RPC that terminated with (StatusCode.UNAVAILABLE, {"created":"@1477577331.417376989","description":"EOF","file":"src/core/lib/iomgr/tcp_posix.c","file_line":235,"grpc_status":14})>
Traceback (most recent call last):
  File "/home/markus/Desktop/channel_test.py", line 40, in <module>
    make_request()
  File "/home/markus/Desktop/channel_test.py", line 30, in make_request
    print('Got successful result: %s' % response.result())
  File "/home/markus/Projects/grpc/_channel.py", line 272, in result
    raise self
_Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.UNAVAILABLE, {"created":"@1477577331.417376989","description":"EOF","file":"src/core/lib/iomgr/tcp_posix.c","file_line":235,"grpc_status":14})>
Sleeping now for 0.2 seconds and then retrying request, which will be successful.
Retrying request now.
check_connectivity_state returned with 0.
Channel changed status to ChannelConnectivity.CONNECTING.
Channel changed status to ChannelConnectivity.READY.
Got successful result: error {
  code: 50001
  err_message: "Status message"
}
```

However, if we reduce the idle time to 300s, the same code succeeds:

```
Making request nr 1.
Channel changed status to ChannelConnectivity.IDLE.
 check_connectivity_state returned with 0.
Channel changed status to ChannelConnectivity.CONNECTING.
Channel changed status to ChannelConnectivity.READY.
Got successful result: error {
  code: 50001
  err_message: "Status message"
}

Sleeping now for 5.0 minutes and then retrying request to produce UNAVAILABLE error.
Making final request.
check_connectivity_state returned with 2.
Got successful result: error {
  code: 50001
  err_message: "Status message"
}
```

Our current theory is that it is related to the different behaviour of the AWS NAT gateway compared to a NAT instance.  This particular link uses the AWS NAT gateway on the exit path from the client VPC.  From the [AWS docs](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/vpc-nat-comparison.html):

> NAT gateway timeout behaviour:
> When a connection times out, a NAT gateway returns an RST packet to any resources behind the NAT gateway that attempt to continue the connection (it does not send a FIN packet).
> 
> NAT instance timeout behaviour:
> When a connection times out, a NAT instance sends a FIN packet to resources behind the NAT instance to close the connection.

**Comment by nathanielmanistaatgoogle** [comment]
@ctiller, might I trouble you for an assessment of this report? Is gRPC doing something wrong, or doing something that could be improved upon, or is this a case in which the observed behavior is as good as it's going to be?

**Comment by ctiller** [comment]
This is going to take some investigation. It's unclear whether we'll need the retry work (slated to begin soon) to resolve it. Let's not pause the 1.1 release.

If it turns out we can fix it easily in that release branch, let's do that, otherwise let's target 1.2.

**Comment by ramurti** [comment]
Hey folks, any news on this? I'm seeing the same problem.

Thanks

**Comment by kpayson64** [comment]
Fixed by #11154

**Comment by fho** [comment]
Could it be that the problem still exist in the PHP client?
We are getting `Sisu\Grpc\Exception\GRPCException: Endpoint read failed` errors on the client on long living and (often) idle connections
